lib/os/cbprintf_nano.c

```diff
@@ -122,175 +69,230 @@ static bool negative(uint_value_type val)
  * @param fmt Format string
  * @param ap Variable parameters
  *
- * @return N/A
+ * @return printed byte count if CONFIG_CBPRINTF_LIBC_SUBSTS is set
  */
 int cbvprintf(cbprintf_cb out, void *ctx, const char *fmt, va_list ap)
 {
 	size_t count = 0;
-	int might_format = 0; /* 1 if encountered a '%' */
-	enum pad_type padding = PAD_NONE;
-	int padlen, min_width = -1;
-	char length_mod = 0;
-
-	/* fmt has already been adjusted if needed */
-	while (*fmt) {
-		if (!might_format) {
-			if (*fmt != '%') {
-				OUTC(*fmt);
-			} else {
-				might_format = 1;
-				min_width = -1;
-				padding = PAD_NONE;
-				length_mod = 0;
+	char buf[DIGITS_BUFLEN];
+	char *prefix, *data;
+	int min_width, precision, data_len;
+	char padding_mode, length_mod, special;
+
+	/* we pre-increment in the loop  afterwards */
+	fmt--;
+
+start:
+	while (*++fmt != '%') {
+		if (*fmt == 0) {
+			return count;
+		}
+		OUTC(*fmt);
+	}
+
+	min_width = -1;
+	precision = -1;
+	prefix = "";
+	padding_mode = 0;
+	length_mod = 0;
+	special = 0;
+
+	for (fmt++ ; ; fmt++) {
+		switch (*fmt) {
+		case 0:
+			return count;
+
+		case '%':
+			OUTC('%');
+			goto start;
+
+		case '-':
+			padding_mode = PAD_TAIL;
+			continue;
+
+		case '.':
+			precision = 0;
+			padding_mode &= ~PAD_ZERO;
+			continue;
+
+		case '0':
+			if (min_width < 0 && precision < 0 && !padding_mode) {
+				padding_mode = PAD_ZERO;
+				continue;
 			}
-		} else {
-			switch (*fmt) {
-			case '-':
-				padding = PAD_SPACE_AFTER;
-				goto still_might_format;
-			case '0':
-				if (min_width < 0 && padding == PAD_NONE) {
-					padding = PAD_ZERO_BEFORE;
-					goto still_might_format;
-				}
-				__fallthrough;
-			case '1':
-			case '2':
-			case '3':
-			case '4':
-			case '5':
-			case '6':
-			case '7':
-			case '8':
-			case '9':
+			__fallthrough;
+
+		case '1'...'9':
```

Comment:
Case ranges are a GNU extension.  Will all required toolchains support it?

Reply:
@tejlmand please comment

Reply:
> Case ranges are a GNU extension. Will all required toolchains support it?

clang also support this syntax, and arcmwdt is clang based so I expect it to work there as well.

But there are also Zephyr users with custom toolchains where we don't know if this is supported.

As there is no functional reason for using this syntax, compared to:
```
case '1':
case '2':
case '3':
...
```
I simply prefer the latter as it's is guaranteed to work in all cases.

Secondly, this code is a good show case why this syntax is dangerous, cause you are supposed to have spaces, like this: `'1' ... '9'`
> Be careful: Write spaces around the ..., for otherwise it may be parsed wrong when you use it with integer values

although this goes for integer values, there is always a risk of developers copy such code and change it, without knowing such details. 
https://gcc.gnu.org/onlinedocs/gcc/Case-Ranges.html